src/common/async/yield_context.h

```diff
@@ -21,25 +21,34 @@
 
 #include "acconfig.h"
 
+#include <optional>
+
 /// optional-like wrapper for a boost::asio::yield_context. operations that take
 /// an optional_yield argument will, when passed a non-empty yield context,
 /// suspend this coroutine instead of the blocking the thread of execution
-class optional_yield {
-  boost::asio::yield_context *y = nullptr;
- public:
-  /// construct with a valid io and yield_context
-  optional_yield(boost::asio::yield_context& y) noexcept : y(&y) {}
 
-  /// type tag to construct an empty object
-  struct empty_t {};
-  optional_yield(empty_t) noexcept {}
+namespace detail { using opt_t = std::optional<boost::asio::yield_context>; }
```

Comment:
rgw always passes optional_yield by value, so we copy these a lot. before this change, that only required copying the pointer. now we have to copy the whole yield_context which, depending on the executor type it wraps, might require reference counting

Reply:
Good catch!! std::optional<> does not model a pointer. 

1) Would it be for our purposes better to write:
shared_ptr<optional_yield> or have optional_yield contain a shared_ptr<> to the yield_context? 

2) I don't see that optional_yield destroys the pointee, what would we use the refcounts for that isn't already happening? (I am only just now getting into the librados calls and eventual asio plumbing this points at, so I may just not understand it yet!) It seems like some of the functions are grabbing the raw pointer out of the optional_yield and passing that down the call stack.

3) I believe that yield_context is routinely passed by value in asio code and contains weak_ptr<> internally, so I expect it's cheap to copy. But, agreed, it's a potential concern. Let's measure this if we do decide that proceeding is worthwhile.

Reply:
> Good catch!! std::optional<> does not model a pointer.

thanks, this is an interesting point about ownership. unlike std::optional, optional_yield doesn't "own" its contents. should it?

as-is, we do risk a dangling pointer if the original yield_context instance goes out of scope. in practice, that seems unlikely - asio::spawn() passes the yield_context as an argument to our coroutine function, so that outlives any local variables. we'd have to do something special to make it dangle, such as:
```c++
void func(boost::asio::yield_context yield)
{
  optional_yield y;
  {
    auto tmp = yield;
    y = optional_yield{tmp};
  }
  otherfunc(y);
}
```
there's also a whole lot of code in rgw, and we have to pass these optional_yield arguments all over the place. i worry that if we do want the safety of shared ownership, that overhead probably will be significant due to the sheer number of function calls that will need to copy it

perhaps we could just do a better job documenting its ownership semantics? changing the name to something like `optional_yield_view` could also help clarify that, but would be a massive diff

Reply:
I don't think we want to touch a bunch of code here. The other good news is that I don't think we break the semantics that the code I see is counting on, but we are offering more options by providing things like std::optional's monadic operations; so, I don't see a downside there.

The overhead is certainly a valid concern. We may be in luck, though: by the time things make it to librados, it looks like we're dealing with asio::yield_context, and in that respect performance should be identical. Let's consider benchmarking it rather than speculating, especially in light of the move construction std::optional<> allows, the yield_context implementation storing a pointer to the actual coroutine context, etc..

Otherwise, maybe some comments just clarifying the differences to the user are sufficient.

Reply:
Okay! Initial measurements show that these are about equal-- there is no penalty to using this patch, insofar as ops/CPU usage is concerned (the PR version may actually be /slightly/ more efficient, but we'll need more measuring to see what the case is there).

Using this benchmark, they appear about identical (grateful to @mkogan1 for the hand here):
time (~/go/bin/hsbench -a 0555b35654ad1656d804 -s h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q== -u http://127.0.0.1:8000/ -z 4K -d -1 -t  $(( $(numactl -N 0 -- nproc) / 1 )) -b $(( $(numactl -N 0 -- nproc) * 2 )) -n 100000 -m cxip -bp b01b- -op 'folder01/stage01_' |& tee hsbench.log)